### PR TITLE
Settling minmax assumptions + tests

### PR DESCRIPTION
**Pull request by pelegm**
I've added some relevant assumption methods for minmax, including tests. It was weird, but I had to add the method `_eval_is_irrational` and `_eval_is_noninteger` - for some reason, there's no proper mixin here, and even if the max was evaluated to be `real=True` and `rational=False`, it didn't conclude that it is `irrational=True` - so I've added it manually.

### CODE REVIEW

File: sympy/functions/elementary/miscellaneous.py [hunk 1]
```diff
@@ -419,6 +419,49 @@ def evalf(self, prec=None, **options):
...
         return fuzzy_and(arg.is_real for arg in self.args)
 
+    def _eval_is_algebraic(self):
+        if all(arg.is_algebraic for arg in self.args):
+            return True
+        if all(arg.is_algebraic is False for arg in self.args):
+            return False
+
+    def _eval_is_even(self):
+        if all(arg.is_even for arg in self.args):
```
Comment (smichr) [inline sympy/functions/elementary/miscellaneous.py]: So it looks like this is waiting for the boilerplate?

**Comment by smichr** [comment]
Something that just occured to me is that all the `_eval_is_foo` methods are doing too much work. The assumption system will query another attribute until it finds the one that is True so, for example, if you define `is_integer` and `is_noninteger` you don't need to include the False test -- the eval routines handle what is True (usually) and the assumption system handles the rest. That being said, I think all the routines can just include your first line in each: `if all(...): return True`.

**Comment by smichr** [comment]
What about making sure it's all covered by generating the routines like this:

```
>>> from sympy.core.assumptions import _assume_defined as d
>>> t = '_eval_is_%s = lambda s: \\\n    all(a.is_%s for a in s.args) or None'
>>> for a in sorted(d):
...  print t % (a, a)
...
_eval_is_algebraic = lambda s: \
    all(a.is_algebraic for a in s.args) or None
_eval_is_antihermitian = lambda s: \
    all(a.is_antihermitian for a in s.args) or None
_eval_is_commutative = lambda s: \
    all(a.is_commutative for a in s.args) or None
_eval_is_complex = lambda s: \
    all(a.is_complex for a in s.args) or None
_eval_is_composite = lambda s: \
    all(a.is_composite for a in s.args) or None
_eval_is_even = lambda s: \
    all(a.is_even for a in s.args) or None
_eval_is_finite = lambda s: \
    all(a.is_finite for a in s.args) or None
_eval_is_hermitian = lambda s: \
    all(a.is_hermitian for a in s.args) or None
_eval_is_imaginary = lambda s: \
    all(a.is_imaginary for a in s.args) or None
_eval_is_infinite = lambda s: \
    all(a.is_infinite for a in s.args) or None
_eval_is_integer = lambda s: \
    all(a.is_integer for a in s.args) or None
_eval_is_irrational = lambda s: \
    all(a.is_irrational for a in s.args) or None
_eval_is_negative = lambda s: \
    all(a.is_negative for a in s.args) or None
_eval_is_noninteger = lambda s: \
    all(a.is_noninteger for a in s.args) or None
_eval_is_nonnegative = lambda s: \
    all(a.is_nonnegative for a in s.args) or None
_eval_is_nonpositive = lambda s: \
    all(a.is_nonpositive for a in s.args) or None
_eval_is_nonzero = lambda s: \
    all(a.is_nonzero for a in s.args) or None
_eval_is_odd = lambda s: \
    all(a.is_odd for a in s.args) or None
_eval_is_polar = lambda s: \
    all(a.is_polar for a in s.args) or None
_eval_is_positive = lambda s: \
    all(a.is_positive for a in s.args) or None
_eval_is_prime = lambda s: \
    all(a.is_prime for a in s.args) or None
_eval_is_rational = lambda s: \
    all(a.is_rational for a in s.args) or None
_eval_is_real = lambda s: \
    all(a.is_real for a in s.args) or None
_eval_is_transcendental = lambda s: \
    all(a.is_transcendental for a in s.args) or None
_eval_is_zero = lambda s: \
    all(a.is_zero for a in s.args) or None
```

**Comment by smichr** [comment]
ping @pelegm - are you going to add the above 25 functions defined in terms of lambdas?

**Comment by pelegm** [comment]
Not all of that. For example, I don't think a maximum of two complex
numbers is complex, as it may be undefined as well.

I'll put some more thought into this soon.

**Comment by smichr** [comment]
> I'll put some more thought into this soon.

OK. But another way to approach this is to recognize that *if* one can say which is bigger or smaller (wrt complex elements) then if all elements are complex, the result will be complex. You can let the assumption system and instantiation worry about whether that will ever happen or not, but from the query point of view you can give an answer: True or None.

**Pull request closed without merging**